docs/graphql/manual/actions/codegen/examples/python-flask/index.rst

```diff
@@ -0,0 +1,387 @@
+
+.. code-block:: python
+
+  import os
+  import jwt
+
+  # Try to get the secret from ENV, else fallback to provided string
+  HASURA_JWT_SECRET = os.getenv("HASURA_GRAPHQL_JWT_SECRET", "a-very-secret-secret")
+
+  # ROLE LOGIC FOR DEMO PURPOSES ONLY
+  # NOT AT ALL SUITABLE FOR A REAL APP
+  def generate_token(user) -> str:
+      """
+      Generates a JWT compliant with the Hasura spec, given a User object with field "id"
+      """
+      user_roles = ["user"]
+      admin_roles = ["user", "admin"]
+      is_admin = user["email"] == "admin@site.com"
+      payload = {
+          "https://hasura.io/jwt/claims": {
+              "x-hasura-allowed-roles": admin_roles if is_admin else user_roles,
+              "x-hasura-default-role": "admin" if is_admin else "user",
+              "x-hasura-user-id": user["id"],
+          }
+      }
+      token = jwt.encode(payload, HASURA_JWT_SECRET, "HS256")
+      return token.decode("utf-8")
+
+  def rehash_and_save_password_if_needed(user, plaintext_password):
+      """
+      Whenever your Argon2 parameters – or argon2-cffi’s defaults! – 
+      change, you should rehash your passwords at the next opportunity.
+      The common approach is to do that whenever a user logs in, since 
+      that should be the only time when you have access to the cleartext password.
+      Therefore it’s best practice to check – and if necessary rehash –
+      passwords after each successful authentication.
+      """
+      if Password.check_needs_rehash(user["password"]):
+          client.update_password(user["id"], Password.hash(plaintext_password))
+
+And finally, ``client.update_password()``:
+
+.. code-block:: python
+
+  update_password = lambda self, id, password: self.run_query(
+      """
+          mutation UpdatePassword($id: Int!, $password: String!) {
+              update_user_by_pk(pk_columns: {id: $id}, _set: {password: $password}) {
+                  password
+              }
+          }
+      """,
+      {"id": id, "password": password},
+  )
+
+Testing out the handler routes
+------------------------------
+
+Now, if we send a request to ``/login`` using our email and password from ``/signup``, we should successfully get a signed JWT, and decoding the JWT should return the correct information:
+
+.. thumbnail:: ../../../../../../img/graphql/manual/actions/python-request-login.png  
+        :width: 90%
+        :alt: Python flask request login
+        :class: no-shadow
+
+.. thumbnail:: ../../../../../../img/graphql/manual/actions/python-flask-jwt-decode.png
+        :width: 80%
+        :alt: Python flask jwt decode
+        :class: no-shadow
+
+Calling the finished actions
+----------------------------
+
+Now we can call our finished Action from Hasura's API, and validate our responses:
+
+Signup Mutation
+^^^^^^^^^^^^^^^
+
+.. thumbnail:: ../../../../../../img/graphql/manual/actions/signup-mutation.png
```

Comment:
Instead of an image, can we show this in GraphiQL style, as in this example: https://hasura.io/docs/1.0/graphql/manual/queries/simple-object-queries.html#fetch-list-of-objects. Then users can copy paste the signup requests.